actors/migration/miner_test.go

```diff
@@ -0,0 +1,492 @@
+		// The sector should be scheduled here, but with faulty power.
+		// It's wrongly here with active power, because fault declaration didn't find it.
+		advancedExpiration: {
+			OnTimeSectors: bitfield.NewFromSet([]uint64{uint64(sectorNo)}),
+			EarlySectors:  bitfield.New(),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   oneSectorPower, // Power should be faulty
+			FaultyPower:   miner0.NewPowerPairZero(),
+		},
+		// The sector should not be scheduled here at all.
+		// It's wrongly here because fault declaration didn't find it at advancedExpiration.
+		faultExpiration: {
+			OnTimeSectors: bitfield.New(),
+			EarlySectors:  bitfield.NewFromSet([]uint64{uint64(sectorNo)}),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   miner0.NewPowerPairZero(),
+			FaultyPower:   oneSectorPower,
+		},
+		// The non-rescheduled sector is still here.
+		// This wrongly has lost power twice.
+		expectedExpiration: {
+			OnTimeSectors: bitfield.NewFromSet([]uint64{uint64(otherSectorNo)}),
+			EarlySectors:  bitfield.New(),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   miner0.NewPowerPairZero(), // Should have oneSectorPower for the other sector
+			FaultyPower:   miner0.NewPowerPairZero(),
+		},
+	})
+
+	phead, err = store.Put(ctx, pst)
+	require.NoError(t, err)
+	setActorState(t, tree, builtin0.StoragePowerActorAddr, builtin0.StoragePowerActorCodeID, phead)
+
+	mhead, err = store.Put(ctx, st)
+	require.NoError(t, err)
+	setActorState(t, tree, maddr, builtin0.StorageMinerActorCodeID, mhead)
+
+	// TODO
+	// - simulate popping the first item off the expiration queue and terminating the sector, which will be
+	//   the more common case we expect to see in the actual state.
+
+	//
+	// Run migration to repair the state
+	//
+	{
+		mm := minerMigrator{}
+
+		migrationEpoch := rescheduleAndFaultEpoch + (10 * builtin.EpochsInDay)
+		newHead, _, err := mm.MigrateState(ctx, store, mhead, MigrationInfo{
+			address:    maddr,
+			balance:    big.Zero(),
+			priorEpoch: migrationEpoch-1,
+			powerUpdates: &PowerUpdates{
+				claims: make(map[addr.Address]power0.Claim),
+				crons:  make(map[abi.ChainEpoch][]power0.CronEvent),
+			},
+		})
+		require.NoError(t, err)
+		require.False(t, newHead.Equals(mhead))
+
+		var st miner2.State
+		require.NoError(t, store.Get(ctx, newHead, &st))
+
+		// Check general state invariants.
+		_, msgs, err := miner2.CheckStateInvariants(&st, store)
+		require.NoError(t, err)
+		assert.True(t, msgs.IsEmpty(), strings.Join(msgs.Messages(), "\n"))
+
+		// Check specifics.
+		_, deadline, partition := loadDeadlineAndPartition2(t, &st, store, dlIdx, pIdx)
+		assertDeadlineExpirationQueue2(t, store, deadline, pIdx, map[abi.ChainEpoch][]uint64{
+			advancedExpiration: {pIdx},
+			faultExpiration:    {pIdx},
+			expectedExpiration: {pIdx},
+		})
+
+		assertPartitionExpirationQueue2(t, store, partition, map[abi.ChainEpoch]*miner2.ExpirationSet{
+			// The sector is  scheduled here, with faulty power.
+			advancedExpiration: {
```

Comment:
This describes the case where the migration is run before the advancedExpiration expires. Given the relatively short window between upgrade and expiry, we're unlikely to find this in the wild. If the advanced expiration has expired, we also expect inaccuracies in partition and power table that need to be corrected, but this is still a good demonstration of the problem and evidence that the migration is working.